File: app/src/main/java/org/watsi/uhp/fragments/EnrollNewbornPhotoFragment.java
package org.watsi.uhp.fragments;

import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import org.watsi.domain.entities.IdentificationEvent;
import org.watsi.domain.entities.Member;
import org.watsi.domain.entities.Photo;
import org.watsi.domain.repositories.MemberRepository;
import org.watsi.uhp.R;
import org.watsi.uhp.managers.LegacyNavigationManager;

import java.io.IOException;

import javax.inject.Inject;

public class EnrollNewbornPhotoFragment extends PhotoFragment<Member> {

    @Inject MemberRepository memberRepository;

    @Override
    int getTitleLabelId() {
        return R.string.enroll_newborn_photo_label;
    }

    @Override
    int getFragmentLayoutId() {
        return R.layout.fragment_capture_photo;
    }

    @Override
    public boolean isFirstStep() {
        return false;
    }

    @Override
    public void nextStep() {
        memberRepository.save(mSyncableModel);
        IdentificationEvent idEvent = (IdentificationEvent) getArguments().getSerializable(LegacyNavigationManager.IDENTIFICATION_EVENT_BUNDLE_FIELD);
        getNavigationManager().setCheckInMemberDetailFragmentAfterEnrollNewborn(mSyncableModel, idEvent);
        Toast.makeText(getContext(), "Enrollment completed", Toast.LENGTH_LONG).show();
    }

    @Override
    void handleSetupFailure() {
        getNavigationManager().setMemberDetailFragment(mSyncableModel);
        Toast.makeText(getContext(), R.string.generic_error_message, Toast.LENGTH_LONG).show();
    }

    @Override
    void additionalSetup(View view) {
        Button savePhotoBtn = ((Button) view.findViewById(R.id.save_button));
        savePhotoBtn.setText(R.string.save_btn_label);

        mSaveBtn.setEnabled(false);
    }

    @Override
    void onPhotoCaptured(Photo photo) throws IOException {
        mSyncableModel.setLocalMemberPhoto(photo);
        mSaveBtn.setEnabled(true);
    }
}

File: app/src/main/java/org/watsi/uhp/fragments/EnrollNewbornPhotoFragment.kt
package org.watsi.uhp.fragments

import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import dagger.android.support.DaggerFragment
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.fragment_capture_photo.photo
import kotlinx.android.synthetic.main.fragment_capture_photo.photo_btn
import kotlinx.android.synthetic.main.fragment_capture_photo.save_button
import org.threeten.bp.Clock
import org.watsi.device.managers.Logger
import org.watsi.domain.entities.Member
import org.watsi.domain.repositories.MemberRepository
import org.watsi.domain.repositories.PhotoRepository
import org.watsi.domain.usecases.CreateMemberUseCase
import org.watsi.uhp.R
import org.watsi.uhp.activities.SavePhotoActivity
import org.watsi.uhp.managers.NavigationManager
import java.util.UUID
import javax.inject.Inject

class EnrollNewbornPhotoFragment : DaggerFragment() {

    @Inject lateinit var clock: Clock
    @Inject lateinit var navigationManager: NavigationManager
    @Inject lateinit var photoRepository: PhotoRepository
    @Inject lateinit var memberRepository: MemberRepository
    @Inject lateinit var createMemberUseCase: CreateMemberUseCase
    @Inject lateinit var logger: Logger

    private var photoIds: Pair<UUID, UUID>? = null

    companion object {
        const val CAPTURE_PHOTO_INTENT = 1
        const val PARAM_MEMBER = "member"

        fun forNewborn(member: Member): EnrollNewbornPhotoFragment {
            val fragment = EnrollNewbornPhotoFragment()
            fragment.arguments = Bundle().apply {
                putSerializable(PARAM_MEMBER, member)
            }
            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        activity.setTitle(R.string.enroll_newborn_photo_label)
        return inflater?.inflate(R.layout.fragment_capture_photo, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        photo_btn.setOnClickListener {
            startActivityForResult(Intent(activity, SavePhotoActivity::class.java), CAPTURE_PHOTO_INTENT)
        }

        save_button.setOnClickListener {
            val newborn = arguments.getSerializable(PARAM_MEMBER) as Member
            createMemberUseCase.execute(newborn.copy(photoId = photoIds?.first))
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({
                navigationManager.popTo(CurrentPatientsFragment())
                Toast.makeText(activity, "Enrollment completed", Toast.LENGTH_LONG).show()
            }, {
                // TODO: handle error
            })
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val (photoIds, error) = SavePhotoActivity.parseResult(resultCode, data, logger)
        photoIds?.let {
            this.photoIds = it
            // TODO: use PhotoLoader
            photoRepository.find(it.second).observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ thumbnailPhoto ->
                        val thumbnailBitmap = BitmapFactory.decodeByteArray(
                                thumbnailPhoto.bytes, 0, thumbnailPhoto.bytes.size)
                        photo.setImageBitmap(thumbnailBitmap)
                    }, {
                        // TODO: handle error
                    })
        }
        error?.let {
            // TODO: handle error
        }
    }
}
